Add release notes for the 0.8.1 release (#1468)
scalameta/metals: website/blog/2020-02-26-cobalt.md

large number of bug fixes and some new features.

<table>
<tbody>
  <tr>
    <td>Commits since last release</td>
    <td align="center">313</td>
  </tr>
  <tr>
    <td>Merged PRs</td>
    <td align="center">100</td>
  </tr>
    <tr>
    <td>Contributors</td>
    <td align="center">19</td>
  </tr>
  <tr>
    <td>Closed issues</td>
    <td align="center">52</td>
  </tr>
  <tr>
    <td>New features</td>
    <td align="center">2</td>
  </tr>
</tbody>
</table>

For full details: https://github.com/scalameta/metals/milestone/19?closed=1

Metals is a language server for Scala that works with VS Code, Vim, Emacs,
Sublime Text, Atom and Eclipse. Metals is developed at the
[Scala Center](https://scala.epfl.ch/) and [VirtusLab](https://virtuslab.com)
with help from [Lunatech](https://lunatech.com) along with contributors from the
community.

## TL;DR

Check out [https://scalameta.org/metals/](https://scalameta.org/metals/) and
give Metals a try!

- insert missing abstract members
- new file provider
- enable rename preview for Visual Studio Code
- debug and run working in Emacs and soon in Vim

## Insert missing abstract members

Thanks to yet again amazing work by

> [tanishiking](https://github.com/tanishiking) we now support the quick fix code

Reviewer: ```suggestion
[tanishiking](https://github.com/tanishiking), we now support the quick fix code
```

action for implementing abstract class members. This code action is available in

> case of errors and works in the same principle as the recent completions. It

Reviewer: I think you can simple this and just say way. "works int he same principle" doesn't read great.
```suggestion
case of errors and works in the same way as the recent "Implement all members" completions. It
```

will add any missing abstract members at the top of the class with the default
implementation of ???.

![action](https://user-images.githubusercontent.com/9353584/73630120-8b0bc880-4698-11ea-8370-32b1fb2f86fd.gif)

## New file provider

> In the last release we added a support for worksheets, which are an easy way to

Reviewer: ```suggestion
In the last release we added support for worksheets, which are an easy way to
```

quickly evaluate some code. To create a new worksheet you need to create a file
with the extension `.worksheet.sc`, which was not that easy to figure out. To
fix that situation, thanks to
[alekseiAlefirov](https://github.com/alekseiAlefirov), we now have a menu to
create different types of new files including classes, traits, object, package
objects and of course worksheets. At this time it's only available in Visual

> Studio Code extension, but will soon be included also in the coc-metals vim

Reviewer: This is already merged into `coc-metals`, so it will work after the release when versioned :)

Reviewer: https://github.com/scalameta/coc-metals/pull/92

extension.

![new-file-provider](https://user-images.githubusercontent.com/10850363/73563467-adfe6880-445d-11ea-89f6-a9a6398034da.gif)

## Enable rename preview for Visual Studio Code

> Rename previews were in the last Visual Studio Code release, which enable users

Reviewer: Thoughts on maybe linking to it in case people aren't sure what we are referring to?

https://code.visualstudio.com/updates/v1_42#_rename-preview

to see what will be changed when a rename is executed. Due to optimization and
UX reasons Metals renamed symbols in closed files in the background without
informing the editor. However, this also causes the previews to be misleading.

To fix that situation we now send all the files to VS Code, which then is able
to display the preview. We do it, however, for up to 300 files, which threshold
was experimentally estimated. When the number of files reaches over that number,

> we revert to the old behavior of not opening closed files. The reason for it is

Reviewer: ```suggestion
we revert to the old behavior of not opening closed files. The reason for this is
```

> that in some cases for larger numbers files the editor could hang and become

Reviewer: ```suggestion
that in some cases for larger numbers of files the editor could hang and become
```

unresponsive.

## Debug in Vim and Emacs

TODO

## Other changes

> This release includes most fixes and minor features, which is due to the scope

Reviewer: ```suggestion
This release includes mostly fixes and minor features, which is due to the scope
```

of the last one. We concentrated last month on stability and making sure
everything is up to scratch. Those minor changes included:

- use bloop.export-jar-classifiers property when importing the build from sbt
- ignore bad compiler options passed by users in order not to break completions

> - add an override for build tool Bloop plugin version and promote it to LSP

Reviewer: ```suggestion
- add an override for the Bloop plugin version and promote it to LSP
```

  </tr>
  <tr>
    <td>Merged PRs</td>
    <td align="center">100</td>
  </tr>
    <tr>
    <td>Contributors</td>
    <td align="center">19</td>
  </tr>
  <tr>
    <td>Closed issues</td>
    <td align="center">52</td>
  </tr>
  <tr>
    <td>New features</td>
    <td align="center">2</td>
  </tr>
</tbody>
</table>

For full details: https://github.com/scalameta/metals/milestone/19?closed=1

Metals is a language server for Scala that works with VS Code, Vim, Emacs,
Sublime Text, Atom and Eclipse. Metals is developed at the
[Scala Center](https://scala.epfl.ch/) and [VirtusLab](https://virtuslab.com)
with help from [Lunatech](https://lunatech.com) along with contributors from the
community.

## TL;DR

Check out [https://scalameta.org/metals/](https://scalameta.org/metals/) and
give Metals a try!

- insert missing abstract members
- new file provider
- enable rename preview for Visual Studio Code
- debug and run working in Emacs

## Insert missing abstract members

Thanks to yet again amazing work by
[tanishiking](https://github.com/tanishiking), we now support the quick fix code
action for implementing abstract class members. This code action is available in

> case of errors and works in the same way as the recent "Implement all members"

Reviewer 2: It might be helpful to add a link to https://scalameta.org/metals/blog/#completion-to-add-all-abstract-members on `recent "Implement all members" completion` for people who are not familiar with the 0.8.0 release.

Author: Good idea! Thanks!

completion. It will add any missing abstract members at the top of the class
with the default implementation of ???.

> ![action](https://user-images.githubusercontent.com/9353584/73630120-8b0bc880-4698-11ea-8370-32b1fb2f86fd.gif)

Reviewer 2: Oops. Sorry again, this behavior in this screen-record is a bit different from the real (we add newlines between each member). I'll take a new one!

Reviewer 2: Just re-took a screen-record with metals 0.8.1, can you use this one instead? :)

![action-take2](https://user-images.githubusercontent.com/9353584/75405616-81047f00-5951-11ea-9972-a12e25b1746a.gif)

Author: Right, I forgot about the newline. Could have made a new gif, but was plain lazy ech.

## New file provider

In the last release we added support for worksheets, which are an easy way to
quickly evaluate some code. To create a new worksheet you need to create a file
with the extension `.worksheet.sc`, which was not that easy to figure out. To
fix that situation, thanks to
[alekseiAlefirov](https://github.com/alekseiAlefirov), we now have a menu to
create different types of new files including classes, traits, object, package

> objects and of course worksheets. At this time it's only available in Visual

Reviewer: ```suggestion
objects and of course worksheets. At this time it's only available in the Visual
```

Studio Code and `coc-metals` extensions.

![new-file-provider](https://user-images.githubusercontent.com/10850363/73563467-adfe6880-445d-11ea-89f6-a9a6398034da.gif)

## Enable rename preview for Visual Studio Code

Rename previews were in the last Visual Studio Code
[release](https://code.visualstudio.com/updates/v1_42#_rename-preview), which
enable users to see what will be changed when a rename is executed. Due to
optimization and UX reasons Metals renamed symbols in closed files in the
background without informing the editor. However, this also causes the previews
to be misleading.

To fix that situation we now send all the files to VS Code, which then is able
to display the preview. We do it, however, for up to 300 files, which threshold
was experimentally estimated. When the number of files reaches over that number,
we revert to the old behavior of not opening closed files. The reason for this
is that in some cases for larger numbers of files the editor could hang and
become unresponsive.

## Debug in Emacs

In the last release run and debug support was only available for Visual Studio
Code, but thanks to amazing work by the contributors it is now possible to use

> it in Emacs via lsp-mode's metals settings and dap-mode.

Reviewer: ```suggestion
it in Emacs via [lsp-mode's](https://github.com/emacs-lsp/lsp-mode) metals settings and dap-mode.
```

## Other changes

This release includes mostly fixes and minor features, which is due to the scope

> of the last one. We concentrated last month on stability and making sure

Reviewer: ```suggestion
of the last one. We concentrated this last month on stability and making sure
```
